### PR TITLE
[ENH] faster opt by changing call to extract_constant and get_scalar_constant_value

### LINKED ISSUE #4004: Verify all call to extract_constant() and get_scalar_constant_value()
If they call is used in an optimization and don't pass the parameter only_process_constants=True, check if that optimization is registered only in Equilibrium optimization that include constant_folding.

If so, add this parameter. It isn't needed and this will make compilation faster by not traversing the graph when not needed.

the canonizer, stabilize and specialize optimizer include the constant_folding optimization. If you find one that is registered in another optimization, report it. Maybe we can include the constant_folding optimization in it.

done partially in https://github.com/Theano/Theano/pull/3974

### PR DESCRIPTION
**Pull request by caglar**
Started adding the only_process_constants=True argument to all extract_constant() and get_scalar_constant_value() calls.

fix gh-4004

### CODE REVIEW

File: theano/compile/tests/test_debugprint.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import theano
```
Comment (nouiz) [inline theano/compile/tests/test_debugprint.py]: Why did you add this file and this test? This is normally tested in theano/tests/test_printing.py

File: theano/sandbox/cuda/check.py [hunk 2]
```diff
@@ -0,0 +1 @@
+import theano; import numpy; a = theano.shared(numpy.zeros(100000).astype('float32'))
```
Comment (nouiz) [inline theano/sandbox/cuda/check.py]: why this file?

File: theano/sandbox/cuda/test_gpuarray.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
+from theano.sandbox.cuda.cula import gpu_solve
```
Comment (nouiz) [inline theano/sandbox/cuda/test_gpuarray.py]: why this file?

**Comment by caglar** [comment]
@nouiz those files were pushed to the repo by mistake. I have just removed them.

File: theano/tensor/opt.py [hunk 4]
```diff
@@ -991,7 +991,7 @@ def shape_ir(self, i, r):
             # Do not call make_node for test_value
             s = Shape_i(i)(r)
             try:
-                s = get_scalar_constant_value(s)
```
Comment (nouiz) [inline theano/tensor/opt.py]: don't do it here.

File: theano/tensor/opt.py [hunk 5]
```diff
@@ -1061,7 +1061,7 @@ def unpack(self, s_i):
             assert len(idx) == 1
             idx = idx[0]
             try:
-                i = get_scalar_constant_value(idx)
```
Comment (nouiz) [inline theano/tensor/opt.py]: idem

File: theano/tensor/opt.py [hunk 6]
```diff
@@ -1120,7 +1120,7 @@ def set_shape(self, r, s, override=False):
                         # But we never timed this speed optimization!
                         self.lscalar_one.equals(shape_vars[i]) or
                         self.lscalar_one.equals(
-                            T.extract_constant(shape_vars[i]))
```
Comment (nouiz) [inline theano/tensor/opt.py]: idem

File: theano/tensor/opt.py [hunk 7]
```diff
@@ -1222,7 +1222,8 @@ def set_shape_i(self, r, i, s_i):
                     # The two following comparison are a speed optimization
                     # But we never timed this speed optimization!
                     self.lscalar_one.equals(new_shape[idx]) or
-                    self.lscalar_one.equals(T.extract_constant(new_shape[idx]))
+                    self.lscalar_one.equals(T.extract_constant(new_shape[idx],
+                                            only_process_constants=True))
```
Comment (nouiz) [inline theano/tensor/opt.py]: idem

File: theano/tensor/opt.py [hunk 8]
```diff
@@ -4957,7 +4971,7 @@ def local_reduce_join(node):
             node.inputs[0].owner and
             isinstance(node.inputs[0].owner.op, T.Join)):
         join = node.inputs[0].owner
-        if T.extract_constant(join.inputs[0]) != 0:
+        if T.extract_constant(join.inputs[0], only_process_constants=True) != 0:
```
Comment (nouiz) [inline theano/tensor/opt.py]: This one is also registered in register_uncanonicalize. Add this around line 5825 in this file:

register_uncanonicalize(topo_constant_folding, 'fast_compile', final_opt=True)

Comment (caglar) [inline theano/tensor/opt.py]: OK @nouiz did the changes that you have suggested.

**Comment by nouiz** [comment]
You didn't merge my PR to your PR with the other fix. I'm killing travis in your new commit.

**Comment by caglar** [comment]
@nouiz  thanks for the notice. I have just realized your PR. For some reason I didn't received any e-mail about it.

**Comment by caglar** [comment]
@nouiz I have added the latest changes that we have discussed and rebased the code.

File: theano/tensor/opt.py [hunk 9]
```diff
@@ -126,15 +126,15 @@ def merge_broadcastables(broadcastables):
...
+def scalarconsts_rest(inputs, elemwise=True, only_process_constants=False):
     """Partition a list of variables into two kinds:
     scalar constants, and the rest."""
     consts = []
     origconsts = []
     nonconsts = []
     for i in inputs:
         try:
-            v = get_scalar_constant_value(i, only_process_constants=True)
+            v = get_scalar_constant_value(i, only_process_constants=only_process_constants)
```
Comment (nouiz) [inline theano/tensor/opt.py]: pass the elemwise kwargs here

File: theano/tensor/nnet/sigm.py [hunk 10]
```diff
@@ -521,7 +522,7 @@ def is_mul(var):
         return None
 
 
-def partition_num_or_denom(r, f):
+def partition_num_or_denom(r, f, **kwargs):
```
Comment (nouiz) [inline theano/tensor/nnet/sigm.py]: this and the change on line 536 didn't get used, so revert them.

**Comment by nouiz** [comment]
killing travis as a few changes are needed.

File: theano/tensor/nnet/sigm.py (hunk 10 above)
Comment (caglar) [inline theano/tensor/nnet/sigm.py]: ok thanks @nouiz  fixed that.

File: theano/tensor/opt.py [hunk 11]
```diff
@@ -126,15 +126,15 @@ def merge_broadcastables(broadcastables):
...
+def scalarconsts_rest(inputs, elemwise=True, only_process_constants=False):
     """Partition a list of variables into two kinds:
     scalar constants, and the rest."""
     consts = []
     origconsts = []
     nonconsts = []
     for i in inputs:
         try:
-            v = get_scalar_constant_value(i)
+            v = get_scalar_constant_value(i, only_process_constants=only_process_constants)
```
Comment (nouiz) [inline theano/tensor/opt.py]: don't forget to pass elemwise here.

File: theano/tensor/opt.py [hunk 12]
```diff
@@ -2317,7 +2319,8 @@ def local_upcast_elemwise_constant_inputs(node):
                 else:
                     try:
                         # works only for scalars
-                        cval_i = get_scalar_constant_value(i, elemwise=False)
+                        cval_i = get_scalar_constant_value(i, elemwise=False,
```
Comment (nouiz) [inline theano/tensor/opt.py]: here you can remove elemwise=False. when we process only constant, this parameter isn't used.

File: theano/tensor/opt.py [hunk 13]
```diff
@@ -3727,8 +3733,10 @@ def local_useless_switch(node):
     """
     if (isinstance(node.op, T.Elemwise) and
             isinstance(node.op.scalar_op, scalar.basic.Switch)):
-        cond = T.extract_constant(node.inputs[0], elemwise=False)
-        if type(cond) is numpy.ndarray and cond.ndim == 0:
+        cond = T.extract_constant(node.inputs[0], elemwise=False,
```
Comment (nouiz) [inline theano/tensor/opt.py]: remove the elemwise parameter

**Comment by nouiz** [comment]
I finished my round of review. Just a few small things. Killing travis

File: theano/tensor/opt.py (hunk 11 above)
Comment (caglar) [inline theano/tensor/opt.py]: ok done.

File: theano/tensor/opt.py (hunk 12 above)
Comment (caglar) [inline theano/tensor/opt.py]: ok, done.

File: theano/tensor/opt.py (hunk 13 above)
Comment (caglar) [inline theano/tensor/opt.py]: done.

**Comment by caglar** [comment]
ok thanks @nouiz finished those changes.

**Comment by nouiz** [comment]
There is this test failing:

```
======================================================================

FAIL: theano.tensor.tests.test_opt.test_local_useless_slice

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/home/travis/miniconda2/envs/pyenv/lib/python2.6/site-packages/nose/case.py", line 197, in runTest

    self.test(*self.arg)

  File "/home/travis/build/Theano/Theano/theano/tensor/tests/test_opt.py", line 1655, in test_local_useless_slice

    assert check_stack_trace(f_unopt, ops_to_check='all')

AssertionError

----------------------------------------------------------------------
```

**Comment by caglar** [comment]
@nouiz this test does not fail locally on my computer. Do you have any idea why it fails?

**Comment by nouiz** [comment]
This triggered an problem that is now fixed in https://github.com/Theano/Theano/pull/4617
